src/delfino/main.py

```diff
@@ -83,17 +89,74 @@ def invoke(self, ctx: click.Context) -> Any:
     def get_help(self, *args, **kwargs) -> str:
         help_str = super().get_help(*args, **kwargs)
 
-        if self._hidden_plugins:
+        if self._hidden_commands:
             if logging.root.level == logging.DEBUG:
                 help_str += click.style(
-                    f"\n\nDisabled command{'s' if len(self._hidden_plugins) > 1 else ''}: "
-                    + ", ".join(sorted(self._hidden_plugins))
+                    f"\n\nDisabled command{'s' if len(self._hidden_commands) > 1 else ''}: "
+                    + ", ".join(sorted(self._hidden_commands))
                     + f" (see 'tool.{ENTRY_POINT}.disable_commands' in '{PYPROJECT_TOML_FILENAME}')",
                     fg="white",
                 )
 
         return help_str
 
+    def _get_commands(self) -> Dict[str, click.Command]:
+        if self._commands:
+            return self._commands
+
+        # default
+        self._commands.update(find_commands(self.PRESET_COMMAND_PACKAGE, required=True))
+
+        # external
+        self._commands.update(self.ExternalCommandsLoader(self._hidden_plugin_commands).load())
+
+        # local
+        self._commands.update(find_commands(COMMANDS_DIRECTORY_NAME, required=False))
+        self._commands.update(find_commands("tasks", required=False, new_name=COMMANDS_DIRECTORY_NAME))
+
+        return self._commands
+
+    class ExternalCommandsLoader:
+
+        PLUGIN_GROUP_NAME = "delfino.commands"
+
+        def __init__(self, disabled_commands: Dict[str, Set[str]] = None):
+            self.disabled_commands = disabled_commands or {}
+
+        def load(self) -> Dict[str, click.Command]:
+            _commands: Dict[str, click.Command] = {}
+            for plugin in self._load_plugins():
+                for name, command in self._load_commands_for_plugin(plugin):
+                    if self._is_disabled_command(plugin.name, name):
+                        continue
+                    if _commands.get(name, False):
+                        raise RuntimeError(
+                            f"Command '{name}' is already registered with {plugin.name}-{plugin.version}."
+                        )
+                    _commands[name] = command
+            return _commands
+
+        def _is_disabled_command(self, plugin_name, command_name):
+            return command_name in self.disabled_commands.get(plugin_name, set())
```

Comment:
You don't need this function if you use [`collections.defaultdict`](https://docs.python.org/3/library/collections.html#collections.defaultdict) instead.

```python
from collections import defaultdict

disabled_commands = defaultdict(set)
assert "disabled_command" in disabled_commands["plugin"] == False
```

Reply:
handled by [893d531](https://github.com/radeklat/delfino/pull/32/commits/893d53156382d6a3bb876a347966e92845e69b3c)